Add `Memory` type

oscardssmith opened the pull request:
This is the implementation of https://hackmd.io/NnLXBeoyRymWgPtHYlW7-A?view#New-Builtin-functions

Fixes https://github.com/JuliaLang/julia/issues/24909

vtjnash committed:
implement Memory{T} as a new backend for Array{T}

also makes memory allocation ccalls safer, for catching Serialization
and deepcopy bugs in packages.

See <https://hackmd.io/@vtjnash/rkzazi7an> for the Julep describing this
change.

quinnj commented on base/array.jl:
```diff
@@ -1063,24 +1076,188 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
+    if delta <= offset
+        a.ref = MemoryRef(ref, 1 - delta, false)
+    else
+        @noinline (function()
```
Can someone describe the use of this `@noinline (function() ...)()` pattern here? What benefit it is providing to the compiler?

oscardssmith replied:
this is for outlining the code. Specifically we want there to be a function call here to make the fast-path faster.

quinnj commented on base/boot.jl:
```diff
@@ -52,8 +52,23 @@
+#    data::Ptr{Cvoid}
+#end
+
 #mutable struct Array{T,N} <: DenseArray{T,N}
```
Any idea how many places we're still relying on Array being defined in /src? Would it be possible to convert all those uses to svec or something?

oscardssmith replied:
unfortunately way too many. Specifically IR and AST needs to be manipulated (both grown and shrunk)

quinnj commented on base/boot.jl:
```diff
@@ -471,34 +488,116 @@ const NTuple{N,T} = Tuple{Vararg{T,N}}
+    end
+    return $(Expr(:new, :self, :(ref), :(dims)))
+end))=#
+eval(Core, :(function (self::Type{Array{T,1}})(ref::MemoryRef, dims::Tuple{Int}) where {T}
```
These evals are because.....we're using `Expr(:new, ...)`? And because we don't have normal constructors available for `Array` since it's /src-defined?

oscardssmith replied:
yes

quinnj commented on base/essentials.jl:
```diff
@@ -1,17 +1,22 @@
+length(a::Array{T, 1}) where {T} = getfield(a, :size)[1]
+length(a::Array) = getfield(getfield(getfield(a, :ref), :mem), :length)
+length(a::GenericMemory) = getfield(a, :length)
+const Memory{T} = GenericMemory{:not_atomic, T}
```
Duplicated definition in base/boot.jl?

quinnj commented on base/genericmemory.jl:
```diff
@@ -0,0 +1,223 @@
+## genericmemory.jl: Managed Memory
+
+"""
+    GenericMemory{isatomic, T} <: AbstractVector{T}
```
Just a minor note that the name `GenericMemory` doesn't really make it obvious that this subtypes `AbstractArray`. Perhaps that's intentional (i.e. we don't want people to treat these too strongly as "arrays").

KristofferC replied:
I really think this shouldn't subtype `AbstractVector`. I doubt you want to use a "named" `GenericMemory` as input to all kinds of linear algebra functions etc. This should IMO be a very "basic" type with not many methods defined on it.

KristofferC replied:
> the question is just, what to do with existing uses of AbstractVector?

Nothing? All I am saying is that I don't think the existing generic methods written for `AbstractArray` will be very useful for `GenericMemory`. Even though you can implement `size` and `getindex`, it doesn't mean you *have* to make it an `AbstractArray` because you then opt into things that you might not want. 

Many methods on `AbstractArray` return an `Array` unless you specialize them but it seems that is rarely something you want to happen if you have a `GenericMemory`. So people will start overloading these methods to return a `GenericMemory` etc and this just doesn't seem like something you want to deal with.

Seelengrab replied:
That makes sense, yes. I'm on board with keeping this as its own thing, and having actual array semantics on different types on top of this.

vtjnash replied:
No, this is all very intentional that it is an AbstractVector, not an accident or pun. In fact, it is much more descriptive of a mathematical vector than Vector was, since Vector adds lots of features as "general collection of objects" which are not mathematically-oriented and do not exist for Memory. However, Vector will remain the preferred type for methods to return in most cases, because Vector is more general. That is also intentional and not something that should be changed in most of those algorithms.

Seelengrab replied:
Fair enough - I guess we'll see how much code expects to `push!` on `AbstractVector` or similar, should it leak out.

vtjnash replied:
Most other AbstractArray types do not support `push!`, so I don't see that being an issue

KristofferC replied:
> In fact, it is much more descriptive of a mathematical vector than Vector was, since Vector adds lots of features as "general collection of objects" which are not mathematically-oriented and do not exist for Memory. 

You want to have a basic memory type that is free from even being thought about in the same way as a mathematical vector. If you want you can wrap something on top of `Memory` that you consider an `AbstractVector` but the core memory type should not be passable into `ForwardDiff.gradient` and to `.+` and a bunch of other stuff that doesn't make sense for it.

oscardssmith replied:
the counterpoint is that given that it has a size, and getindex, you will get sensible results out of all of those.

vtjnash replied:
Only to answer the last question, which may answer most of the others, they should be taking a DenseArray already, so no changes are needed for them. (Except that reminds me that this should really be a DenseVector, and not merely an AbstractVector. Usually I don't bother to express that level of granularity in array APIs, so it had slipped my mind.)

KristofferC replied:
> which may answer most of the others

I don't really see how the other questions are answered from this.

mikmoore commented:
Sorry for the bikeshedding, but given that we already have `RefArray` and `RefValue`, `RefMemory` seems more consistent than the proposed `MemoryRef`. Although I'll admit that I like `XyzRef` better than `RefXyz` in a vacuum. In any case, very exciting!

JeffBezanson commented:
> Sorry for the bikeshedding, but given that we already have `RefArray` and `RefValue`, `RefMemory` seems more consistent than the proposed `MemoryRef`. Although I'll admit that I like `XyzRef` better than `RefXyz` in a vacuum. In any case, very exciting!

That's a good point, but I reason that `RefArray` and `RefValue` are not used directly very often and `MemoryRef` serves a slightly different purpose, plus I agree is generally a better name. Although like RefArray a MemoryRef points to an element, its real purpose is to capture a bounds-checked index.

Seelengrab commented:
Some questions that came up while reading the design document - documenting here so they're not lost to the triage hole:

 * How is a `:const` memory going to be initialized by a user? In embedded, it's useful & common to have some global immutable constant that's already initialized at link time.
 * Will the memory model of atomics be documented together with the documentation of this atomic memory? See also https://github.com/JuliaLang/julia/issues/46739
 * Could this support address spaces from LLVM? (I think we talked about this a bit in the context of GPU & embedded, where we'd otherwise have to reimplement a whole bunch of stuff)
 * Can I take a `view` of a `GenericMemory`? (Does that even make sense..?)

oscardssmith commented:
1. this would probably require a special constructor that takes a function that takes an `Int` and produces a `T` (or we could make a copy constructor from a non-const `Memory`
2.  it really should be
4. yes. A 1 element view of a `GenericMemory` is a `MemoryRef`. A multi-element view of a `GenericMemory` is another `GenericMemory` owned by the parent `GenericMemory` (see `jl_genericmemory_slice`)

oscardssmith commented on base/Base.jl:
```diff
@@ -275,24 +288,24 @@ include("set.jl")
+        # if data is in a String, keep it that way
+        # TODO: use jl_gc_expand_string(oldstr, newlen)?
+        str = _string_n(newlen)
+        return (@assume_effects :total !:consistent @ccall jl_string_to_genericmemory(str::Any,)::Memory{UInt8})
```
is this really effect free?

vtjnash replied:
yes, allocation doesn't have side effects

Tokazama commented:
I'm trying to help where I can with review and documentation here, but it's a lot to sift through. Do we still need to bike-shed the resize/grow/extend stuff, or is there something else more helpful you'd like more eyes on?

vtjnash commented:
We are slowly cranking through the todo list items listed in the commit comments (like making sure everything is documented and tested), and making sure all of the tests continue to pass. If people want to help, they are welcome to claim items via Slack for real-time collaboration on them

oscardssmith commented on base/essentials.jl:
```diff
@@ -7,8 +7,9 @@ const Callable = Union{Function,Type}
+length(a::Array{T, 0}) where {T} = 1
 length(a::Array{T, 1}) where {T} = getfield(a, :size)[1]
-length(a::Array) = getfield(getfield(getfield(a, :ref), :mem), :length)
+length(a::Array) where {T} = getfield(getfield(getfield(a, :ref), :mem), :length)
```
what's the `T` here?

oscardssmith commented on base/essentials.jl:
```diff
@@ -7,8 +7,9 @@ const Callable = Union{Function,Type}
 const Bottom = Union{}
 
 # Define minimal array interface here to help code used in macros:
+length(a::Array{T, 0}) where {T} = 1
```
why specialize?

vtjnash committed:
Revert "cleanup boot"

This reverts commit 3ccbee76b9cd28c5b7990ccd37aca9ce36e10370 while we
decide if we want to do this or not, though it may be worthwhile

Tokazama commented on base/genericmemory.jl:
```diff
@@ -0,0 +1,261 @@
+"""
+GenericMemory
+"""
+    Memory{T} == GenericMemory{:not_atomic, T}
```
It would be nice to have a more formal representation of accessibility and mutability. Maybe a formal type that we could attach documentation to and dispatch on, if necessary. I understand that we don't want to overcommit to a hierarchy of types since the meaning could change based on context (annotating a function vs a stored object). But even something like this might be useful.

```julia
baremodule MemoryKinds
    struct Constant end
    baremodule Mutable
        struct NotAtomic end
        struct Unordered end
        struct Monotonic end
        struct Acquire end
        struct Release end
        struct AcquireRelease end
        struct SequentiallyConsistent end
    end
end
```

oscardssmith replied:
so the currently expected options are `not_atomic` and `atomic`. For `atomic` memories, the kind of memory order is a property of the `setindex!`, not of the datatype.

Tokazama replied:
> For `atomic` memories, the kind of memory order is a property of the `setindex!`, not of the datatype.

I'm not currently seeing any code for the `atomic` variant of `setindex!`. Do you mean that the exact order in which memory is accessed is specified when calling `setindex!`, similar to how `unsafe_store!` works?

I thought there was talk of also having a constant variant for that parameter and figured this would provide explicit support in the type for what kind of access is supported. So the atomic variant would be `GenericMemory{Union{Unordered, Monotonic, Acquire, Release, AcquireRelease, SequentiallyConsistent}, T}`, the current implementation would be `GenericMemory{NotAtomic, T}`, and the constant variant would be `GenericMemory{Constant, T}`.

We could also just punt the full specification down the line a bit, but if we do then we should probably remove explicit mention of that field so it's not erroneously considered a stable interface that people should mess with.

oscardssmith replied:
as mentioned in the design doc, most of the code for atomic `GenericMemory` is not written yet. We just need to know what the design will be so we don't discover in the future that we need an extra type parameter that would be breaking to add
> Do you mean that the exact order in which memory is accessed is specified when calling setindex!, similar to how unsafe_store! works?

yes.

Tokazama replied:
Apologies, I wasn't aware that the design doc was still undergoing revisions at this point. Would it be better to comment there since this detail may not make it into this PR?

Seelengrab replied:
The primary argument against making the kind type parameter a proper type was that there hardly is any hierarchy between memory kinds. They are (mostly) mutually exclusive, and so using a symbol for them is usually fine. Not being able to add docstrings to those symbols is a bit of an argument, but something that could be solved by

```
const NotAtomic = :not_atomic
```

too, where the docstring would then be attached to `NotAtomic`. Personally, I dislike this kind of magic-constantness (Ref. https://github.com/JuliaLang/julia/issues/51357) because it hides the fact that this object has a special semantic meaning on the value/type level - you're not supposed to use that object outside of the narrow context of memory kinds in `GenericMemory` after all. 

Maybe this is on example where being able to specify an `isa` relationship in `where` clauses would be beneficial, in order to ensure that not just any random symbol is accepted as a memory kind.

Tokazama replied:
The main benefit really comes from being able to dispatch on kinds (as discussed in the [JuliaAtomics proposal](https://gist.github.com/vtjnash/11b0031f2e2a66c9c24d33e810b34ec0#alternatives-considered)) and test the validity of a kind of memory access.

For pointers and properties/fields, this is mostly a matter of convenience since we could probably do some compiler work to ensure symbols and modifying functions propagate appropriately. However, the array interface is a lot messier since `getindex(kind, a::AbstractArray, inds...)` may have to propagate `kind` through a number of wrapping structures before arriving at the innermost type and being tested for validity.

Seelengrab replied:
The JuliaAtomics manifesto has so many TODOs/things that must have been decided by now due to being required for the implementation that I really don't think pointing to it as a source of truth is helpful (ref. https://github.com/JuliaLang/julia/issues/46739). I also don't see any reference to dispatching on kind types; those are always symbols there.

Tokazama commented:
Currently getting the same build error found here https://buildkite.com/julialang/julia-master/builds/27964#018ab447-4139-4195-b2d2-d58f8ce4b059/466-1210

Tokazama commented on base/Base.jl:
```diff
@@ -186,12 +195,26 @@ include("strings/lazy.jl")
+const DL_LOAD_PATH = String[]
+let os = ccall(:jl_get_UNAME, Any, ())
+    if os === :Darwin || os === :Apple
+        if Base.DARWIN_FRAMEWORK
```
Currently getting the same build error as found here https://buildkite.com/julialang/julia-master/builds/27964#018ab447-4139-4195-b2d2-d58f8ce4b059/466-1210

Tokazama replied:
Currently getting the same build error as found here https://buildkite.com/julialang/julia-master/builds/27964#018ab447-4139-4195-b2d2-d58f8ce4b059/466-1210

mbauman commented on base/Base.jl:
```diff
@@ -6,10 +6,19 @@ using Core.Intrinsics, Core.IR
-const _included_files = Array{Tuple{Module,String},1}(Core.undef, 1)
+# start this big so that we don't have to resize before we have defined how to grow an array
+const _included_files = Array{Tuple{Module,String},1}(Core.undef, 50)
+_included_files.size = (1,)
```
Just a small thing — perhaps we should specialize `getproperty` to make `Array` look fully opaque from the REPL (and require get/set field for this kind of internals muckery).  Given other languages, `arr.size` is a really tempting footgun.

oscardssmith replied:
yeah that's a good point. As nice as it made coding these, they should probably require a `setfield`

Seelengrab replied:
Perhaps a name other than `size` could serve an additional purpose, to further reduce "just access this"? Something like `dims_tuple`?

KristofferC replied:
What does `setfield` has to do with REPL autocompletions? 

And what's wrong with Matt's original suggestion again?

oscardssmith replied:
what I mean is that if we overload `getproperty` (and presumably `setproperty`) you would need `setfield` to modify the size. I was agreeing with the original suggestion.

KristofferC replied:
Isn't it enough to overload `propertynames` which the REPL use for autocomplete?

quinnj commented on base/array.jl:
```diff
@@ -1063,24 +1030,182 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
+        a.ref = MemoryRef(newmem, newoffset+1, false)
+        end)()
+    end
+    return
```
We still need to account for type tag bytes for bitsunion arrays in these `_growX` functions, right? Is that just a TODO still?

oscardssmith replied:
no. That just works. (The Memory keeps track of the type tags too)

vtjnash commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1737072851) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/d13bdc0_vs_0a82b71/report.md).

PallHaraldsson commented:
It benchmarks often much faster 50%+ (even see down to 0.02 which I do not quite believe), but I also see up to almost 8x slower, which seems suspicious (sometimes measuring noise?). [Also surprisingly helping (or hurting) in memory column.]

Should these benchmarks show outliers in both directions at the top, since it would be helpful to not have to scroll through the whole list? I guess only if the noise can be eliminated. Could the list be cut down to a reliable set, and at least show outliers for it? And maybe keep the other tests in a separate category below?

oscardssmith commented:
the 3-8x regressions are real. I deleted a few optimization methods for array `getindex` that I was hoping the compiler was now smart enough to figure out (and which were causing some ambiguity conflicts). We clearly seem to need them still, so now it's time to see if we can add them (and hopefully add them for all `DenseArray` so it's not just `Array` that benefits.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1740642732) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/cf188c3_vs_b74daf5/report.md).

Tokazama commented:
I just noticed there's no direct `Memory`/`MemoryRef`/`GenericMemory` function for constructing an initial instance. Is the current plan to keep direct construction as an internal implementation so there's room for more design in the near future?

oscardssmith commented:
There's `Memory{T}(undef, n)`. Were you thinking about something else?

Tokazama commented:
> There's `Memory{T}(undef, n)`. Were you thinking about something else?

Sorry, I did `methods(Memory)` instead of `methods(Memory{Any})` so I didn't see those. It's not in the `Memory` doc string though, so I thought it might be intentional.

Do we want to create an additional doc string for the `undef` method that's like the one for `Array`

```julia
julia> m = Memory{Float64}(undef, 2)
2-element Memory{Float64}:
 0.0
 0.0

julia> similar(m, 3)
3-element Memory{Float64}:
 0.0
 0.0
 0.0

```

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1749764186) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/b146a70_vs_20a5fa7/report.md).

PallHaraldsson commented:
This seems very good with low fractions down to e.g. 0.16 but even 0.00 which is suspect (100% speedup), and as before sometimes 7x, which was expected, but also this strange:

"broadcast", "mix_scalar_tuple", (10, "scal_tup_x3")]	734.79 (5%) ❌	Inf (1%)

aviatesk commented on stdlib/REPL/src/REPLCompletions.jl:
```diff
@@ -481,7 +481,8 @@ CC.haskey(wvc::CC.WorldView{REPLInterpreterCache}, mi::MethodInstance) = haskey(
 
 # REPLInterpreter is only used for type analysis, so it should disable optimization entirely
-CC.may_optimize(::REPLInterpreter) = false
+# however, since effects are not computed until after optimization, the optimizations cannot be disabled
+CC.may_optimize(::REPLInterpreter) = true
```
```suggestion
CC.may_optimize(::REPLInterpreter) = false
```
to fix the broken test case, because otherwise we may hit this case: https://github.com/JuliaLang/julia/blob/1abafe85b2c51417dc145c90e8dde72485293b6c/base/compiler/abstractinterpretation.jl#L790-L793

While it is true that post-opt analysis can refine effects, I doubt the benefits of that refinement really make up for the cost of going through all those optimization passes. 
Since post-opt analysis refinement doesn't really widen the effects, I think it's fine to cache the effects that come from abstract interpretation.

gbaraldi commented:
@vtjnash Something that we may want to do for the addresspace stuff is to make the size of the pointer parametric on the addresspace, examples being AMGPU private memory (addrspace 5) is 32 bit addressed while the global memory is 64bit addressed.

Seelengrab commented on base/array.jl:
```diff
@@ -1063,24 +1027,182 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
+function _deletebeg!(a::Vector, delta::Integer)
+    delta = Int(delta)
+    len = length(a)
+    0 <= delta <= len || throw(ArgumentError("_deleteat! requires delta in 0:length(a)"))
```
```suggestion
    0 <= delta <= len || throw(ArgumentError("_deletebeg! requires delta in 0:length(a)"))
```

Seelengrab commented on base/array.jl:
```diff
@@ -1063,24 +1027,182 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
+function _deleteend!(a::Vector, delta::Integer)
+    delta = Int(delta)
+    len = length(a)
+    0 <= delta <= len || throw(ArgumentError("_deleteat! requires delta in 0:length(a)"))
```
```suggestion
    0 <= delta <= len || throw(ArgumentError("_deleteend! requires delta in 0:length(a)"))
```

Seelengrab reviewed (changes requested):
Just some error message suggestions, aligning the error message with the function it was called in :)

Tokazama commented on base/genericmemory.jl:
```diff
@@ -0,0 +1,273 @@
+sizeof(a::GenericMemory) = Core.sizeof(a)
+
+# multi arg case will be overwritten later. This is needed for bootstraping
+function isassigned(a::Memory, i::Int)
+    @inline
+    @boundscheck 1 <= i <= length(a) || return false
+    # @boundscheck (i % UInt) - 1 < length(A) || return false
+    return @inbounds memoryref_isassigned(GenericMemoryRef(a, i), :not_atomic, false)
+end
```
As far as I can tell the C methods backing `memoryref_isassigned` aren't used in "src", so we might be able to move this completely to Julia code and clean up some of the the source code. Here's what I was thinking.

```julia
function isassigned(mem::Memory, i::Int)
    @inline
    @boundscheck 1 <= i <= length(mem) || return false
    memT = typeof(mem)
    arrayelem = datatype_arrayelem(memT)
    elsz = datatype_layoutsize(memT)
    # might need to be mad
    firstptr = unsafe_load(unsafe_convert(Ptr{DataTypeLayout}, memT.layout)).firstptr % Int
    isboxed = 1; isunion = 2
    if arrayelem == isboxed
        p = Ptr{Ptr{Cvoid}}(@inbounds pointer(mem))
        t = @_gc_preserve_begin mem
        out = Intrinsics.atomic_pointerref(p, :monotonic) != C_NULL
        @_gc_preserve_end t
    elseif arrayelem != isunion && firstptr >= 0
        p = Ptr{Ptr{Cvoid}}(@inbounds pointer(mem)) + fieldoffset(eltype(memT), firstptr + 1)
        t = @_gc_preserve_begin mem
        out = Intrinsics.atomic_pointerref(p, :monotonic) != C_NULL
        @_gc_preserve_end t
    else
        out = true
    end
    return out
end
```

vtjnash replied:
We could, but that code is less correct than the code it replaces, so no, that is not a good implementation

Tokazama replied:
But could we use a Julia implementation instead? This was just my first quick attempt, so if we can still get performance and correctness in Julia it seems like a big win to get rid of all that extra source code.

oscardssmith replied:
the fact that this necessarily involves casting to `Ptr{Ptr{Cvoid}}` implies to me that this should probably stay on the C side. Keeping the Julia side "normal" is nice.

Tokazama replied:
It's not the only place in Julia code with `Memory` that this is handled, but I do see the merit in being cautious nonetheless since we don't have a fully public API for this kind of pointer.

gbaraldi replied:
This probably should be an intrinsic

Seelengrab commented on base/array.jl:
```diff
@@ -1063,24 +1027,183 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
+function _deletebeg!(a::Vector, delta::Integer)
+    delta = Int(delta)
+    len = length(a)
+    0 <= delta <= len || throw(ArgumentError("_deleteat! requires delta in 0:length(a)"))
```
```suggestion
    0 <= delta <= len || throw(ArgumentError("_deletebeg! requires delta in 0:length(a)"))
```

Seelengrab commented on base/array.jl:
```diff
@@ -1063,24 +1027,183 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
+function _deleteend!(a::Vector, delta::Integer)
+    delta = Int(delta)
+    len = length(a)
+    0 <= delta <= len || throw(ArgumentError("_deleteat! requires delta in 0:length(a)"))
```
```suggestion
    0 <= delta <= len || throw(ArgumentError("_deleteend! requires delta in 0:length(a)"))
```

Seelengrab reviewed (changes requested):
The force push by @vtjnash reverted the error message fix.

vtjnash committed:
try to reduce overallocation of system arrays

These arrays often tend to grow slowly over long periods of time, so it
seems likely better to keep them small to save memory, but not so small
that the first 4 push calls each generate a new allocation.

Overall this seems to make the sysimg slightly bigger currently.

simeonschaub commented on base/array.jl:
```diff
@@ -187,12 +185,15 @@ function vect(X...)
-size(a::Array{<:Any,N}) where {N} = (@inline; ntuple(M -> size(a, M), Val(N))::Dims)
+size(a::Array, d::Integer) = size(a, Int(d)::Int)
+function size(a::Array, d::Int)
+    d < 1 && error("arraysize: dimension out of range")
```
This probably shouldn't refer to `arraysize` anymore at least, but might as well make this a more useful error while we're at it?
```suggestion
    d < 1 && throw(ArgumentError(lazy"size: invalid dimension $d for $(ndims(a))d Array")
```
Or are we worried about introducing more compiler-side complexity in this hot path?

simeonschaub commented on src/array.c:
```diff
@@ -215,10 +215,16 @@ JL_DLLEXPORT void jl_array_grow_end(jl_array_t *a, size_t inc)
+            newmaxsize = oldmaxsize*6/5; // grow by 20%
         if (newmaxsize < reqmaxsize)
             newmaxsize = reqmaxsize;
+        // TODO: round this up to newmaxsize < GC_MAX_SZCLASS ? jl_gc_sizeclasses[jl_gc_szclass(newmaxsize)] : LLT_ALIGN(newmaxsize, 4096), after accounting for the object header (24 bytes)
```
Am I understanding correctly that the growing heuristic is now different on the Julia side from the C API?

vtjnash replied:
yes

oscardssmith replied:
the motivation is that the C resizing only happens for arrays in the compiler where we expect a lot of slow growing arrays, at which point saving memory matters more than saving time.

simeonschaub commented on base/genericmemory.jl:
```diff
@@ -0,0 +1,273 @@
+@eval begin
+function setindex!(A::Memory{T}, x, i1::Int) where {T}
+    val = x isa T ? x : convert(T,x)::T
+    ref = memoryref(memoryref(A), i1, $(Expr(:boundscheck)))
```
Should we perhaps make `$(Expr(:boundscheck))` just `@boundscheck()`? Then we wouldn't always have to use `eval` here

vtjnash replied:
It is not needed outside of these few cases, and boundscheck already has a different meaning

vtjnash commented:
@nanosoldier `runtests()`

vtjnash commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

vtjnash commented:
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1762507607) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/5a9df24_vs_3250804/report.html) is available.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1763406423) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/39bafe9_vs_b1e711e/report.md).

vtjnash commented:
That looks very good. There is only one regression that I don't recognize in that list that likely merits some investigation:
```
["collection", "deletion", ("Vector", "Any", "filter!")]	7.77 (25%) ❌	Inf (1%) ❌
["collection", "deletion", ("Vector", "Any", "filter")]	12.33 (25%) ❌	1.48 (1%) ❌
["collection", "deletion", ("Vector", "Any", "pop!")]	1.50 (25%) ❌	1.00 (1%)
["collection", "deletion", ("Vector", "Int", "filter!")]	1.01 (25%)	Inf (1%) ❌
["collection", "deletion", ("Vector", "Int", "filter")]	2.10 (25%) ❌	1.48 (1%) ❌
["collection", "deletion", ("Vector", "String", "filter!")]	1.29 (25%) ❌	Inf (1%) ❌
["collection", "deletion", ("Vector", "String", "filter")]	1.47 (25%) ❌	1.56 (1%) ❌
```
@nanosoldier `runbenchmarks("collection" && "deletion", vs=":master")`

oscardssmith commented:
Should we also have a PR for OrderedCollections to move them to the new api?

vtjnash commented:
@nanosoldier `runtests(["ModiaBase", "MIRTio", "LightOSM", "PlasmoPlots", "TexTables", "Determinantal", "Agents", "Dualization", "Crispulator", "ProfileVega", "SummarizedExperiments", "NaNStatistics", "EasyConfig", "ApproxFunSingularities", "PointSpreadFunctions", "CIMOptimizer", "Loraine", "LinearMaps", "DynamicSparseArrays", "NaiveNASlib", "TypedPolynomials", "ImageMorphology", "PlantRayTracer", "CxxWrap", "MadNLP", "HurdleDMR", "Ipopt", "MendelPlots", "Petri", "StationXML", "PALEOboxes", "Singular", "TabletWeaving", "Resizing", "Demes", "RenderJay", "PreallocationTools", "CCBlade", "PhaseRec", "StrideArraysCore", "PySA", "SqpSolver", "GLPK", "Groups", "RxEnvironments", "PortfolioAnalytics", "SolverBenchmark", "MimiBRICK", "Anneal", "CitablePhysicalText", "MosekTools", "SCS", "SMLMData", "SparseKmeansFeatureRanking", "PerfChecker", "TimeSeriesClassification", "PowerModelsAnalytics", "NamedArrays", "PeriodicSchurDecompositions", "SeisPDF", "NgSpice", "DisjunctiveProgramming", "OrderedCollections", "FullNetworkSystems", "QuantumControlBase", "sparseQFCA", "Cbc", "Thermodynamics", "MutableArithmetics", "DWaveNeal", "PointCloudRasterizers", "FourierTools", "DelayDiffEq", "SDDP", "HierarchicalUtils", "ONSAS", "GpABC", "Bobby", "DistributedFactorGraphs", "Microbiome", "PowerFlows", "InteractiveDynamics", "PNGFiles", "Muspel", "FastRunningMedian", "PlotlyLight", "SparseIR", "JuDGE", "OptimizedEinsum", "ManifoldDiffEq", "Garlic", "DynamicAxisWarping", "Dictionaries", "CrystalNets", "SumOfSquares", "JLD2", "DataFlowTasks", "DataDrivenAcoustics", "Plasmo", "E4STUtil", "Gaius", "SignalTablesInterface_PyPlot", "KalmanFilterTools", "LinearNoiseApproximation", "ODEProblemLibrary", "DifferentiableCollisions", "DataDrivenDMD", "NestedSamplers", "QUBODrivers", "OSMMakie", "NaiveNASflux", "LocalPoly", "SHTOOLS", "MCPTrajectoryGameSolver", "IsingSolvers", "TMLE", "StatusSwitchingQP", "ModelPredictiveControl", "DelaySSAToolkit", "OMETIFF", "LSODA", "MQLib", "Strided", "CombinatorialSpaces", "ComplexMixtures", "Gadfly", "BSeries", "Alexya", "GModelFitViewer", "GasChromatographySimulator", "OrdinalGWAS", "ImageSegmentation", "BilevelJuMP", "OpenQASM", "MomentClosure", "MTKHelpers", "OSQP", "BayesNets", "ConstructionBase", "VIDA", "ImplicitDifferentiation", "DIVAnd", "DiffEqProblemLibrary", "FractionalSystems", "QuantumAnnealingAnalytics", "UnionCollections", "GeneticsMakie", "Phylo", "ReduceWindows", "GaussianBasis", "CSetAutomorphisms", "Intervals", "Boscia", "RData", "ArviZPythonPlots", "StochasticDelayDiffEq", "KiwiConstraintSolver", "Krylov", "LinearRegressionKit", "GeNIOS", "LineIntegralConvolution", "ParticleFilters", "Attractors", "SignalTablesInterface_WGLMakie", "TrueSkillThroughTime", "ElectronLiquid", "ECOS", "AffineMaps", "CirculatorySystemModels", "LibModbus", "GraphQLGen", "TypeClasses", "GeoAcceleratedArrays", "ClosedLoopReachability", "ChaosTools", "ProtoBuf", "ClimateTools", "AiidaDFTK", "FinEtoolsVoxelMesher", "ConstraintSolver", "AmplNLWriter", "HighVoronoi", "RBNF", "Jabalizer", "Shapefile", "BioFetch", "MCMCChainsStorage", "GenomicAnnotations", "SignalTablesInterface_CairoMakie", "MixedModelsPermutations", "HomalgProject", "BundleMethod", "Modia", "Consensus", "RuntimeGeneratedFunctions", "FiniteDiff", "Flatten", "Distances", "AlgebraicRewriting", "WordCloud", "NeuronBuilder", "EpithelialDynamics1D", "ComplexityMeasures", "QuantumOptics", "Catalyst", "PowerPlots", "SerializationCaches", "SimpleConfig", "GPUArrays", "Scruff", "PDEBase", "LatticeQCD", "ImageIO", "SemiLagrangian", "UnitDiskMapping", "PPTX", "AsyPlots", "MRINavigator", "SymPyPythonCall", "UnitJuMP", "GlobalSensitivity", "vOptGeneric", "VegaLite", "YAML", "Conductor", "HmtArchive", "Controlz", "TranscodingStreams", "VideoIO", "ReactionSensitivity", "ControlSystemsMTK", "MiseEnPage", "MLJModelInterface", "ReactionNetworkImporters", "BandedMatrices", "SMLMFrameConnection", "GeoIP", "NetDecOPF", "ChunkedBase", "IRKGaussLegendre", "SimpleCrop", "JumpProcesses", "GeometricTheoremProver", "PowerSystemsMaps", "ParametricOptInterface", "MinimallyDisruptiveCurves", "StaticArrayInterface", "BranchFlowModel", "JUDI", "LazyAlgebra", "Materials", "ParametricLP", "DifferentiableBackwardEuler", "EMpht", "Nucleotide_Essentials", "SciMLExpectations", "QuantumClifford", "MimiSSPs", "ModeCouplingTheory", "PowerAnalytics", "SDEProblemLibrary", "Sundials", "SurfaceCoverage", "EditorsRepo", "ImageAnnotationsIO", "WaveguideQED", "SCIP", "CropRootBox", "Sunny", "AbstractAlgebra", "ManifoldDiff", "VoronoiFVM", "TrajOptPlots", "FlashWeave", "SentinelArrays", "Elliptic", "Jutul", "Fenrir", "PRONTO", "ManifoldsBase", "Aerosol", "ODE", "Powersense", "VulkanSpec", "DynamicPolynomials", "GMT", "SUNRepresentations", "TerminalUserInterfaces", "CSDP", "FractalDimensions", "CachedFunctions", "JsonGrinder", "PlantGeomTurtle", "PowerNetworkMatrices", "Undefs", "SignalTables", "PointEstimateMethod", "Clp", "Fronts", "DiffusionMap", "Pavito", "PolyJuMP", "BM3DDenoise", "ImageFiltering", "COPT", "HiGHS"])`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1763679856) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/aad7090_vs_258ac46/report.md).

vtjnash commented:
@nanosoldier `runbenchmarks("collection" && "deletion", vs=":master")`

nanosoldier commented:
[Your job](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1764825000) failed.

vtjnash commented:
@nanosoldier `runbenchmarks("collection" && "deletion", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1765245644) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6885284_vs_d5cde52/report.md).

rapus95 commented:
Are there any news or insights for what has happened to the proposal from triage to be able to specify the memory bit width as a parameter of AddrSpace for being able to properly model different memory widths for different memories? (I.e. 32, 48 and 64 bit pointers and memory).

oscardssmith commented:
The answer we came up with is that due to alignment, shrinking the field won't save any space, so it's the responsibility of the code in the address space to ignore the bits it doesn't care about.

rapus95 commented:
@oscardssmith But given that the AddrSpace type isn't usually named explicitly wherever the CPU space is used, but only for extending to different architectures, what's the drawback of just plugging it in to have the nominal and conceptual distinctiveness? Aka designing for descriptiveness instead of shortness when length won't matter for "default" code (since there it'll be abbreviated by the const CPU) anyway. In particular, when it's also entirely hidden from the c and unsafe implementations level since these will by default only be defined for CPU which already would refer to the width of Int. Anything like GPU support also will only define for their own AddrSpace probably also using a const. So it most likely will make neither coding nor reading more difficult or cumbersome.

Seelengrab commented:
> The answer we came up with is that due to alignment, shrinking the field won't save any space, so it's the responsibility of the code required for the address space to ignore the bits it doesn't care about.

I'm not sure I follow the logic - are you saying because codegen (currently) only cares about its own architecture, its alignment is forced on any other targets anyway and thus this doesn't matter?

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1763690735) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/abeb0f4_vs_3250804/report.html) is available.

PallHaraldsson commented:
Way too many packages are failing e.g. JLD2 with:
>ERROR: LoadError: Type Module does not have a definite size.

https://github.com/JuliaLang/julia/blob/d42415ae79613970ca366c207d478988b31332e6/src/builtins.c#L521

I think at first you were only adding a (Memory) type, now replacing Array which was the point. Does it make sense to replace gradually? I'm thinking if I can help, and were to begin, this seems like a lot of new code... I suppose Module uses Array, and then now new Memory, but it could use the old Array type that worked?

vtjnash committed:
replace unsafe_convert methods with safe cconvert ones

This is the correct way to define most conversions, since the unsafe
logic should typically be isolated to a few areas, and everywhere else
just defines conversions to it. This ensures the root is preserved even
if the user later makes unexpected changes to the array object (although
the length might have later become inconsistent at that point, so it is
not a guaranteed fix).

vtjnash committed:
ensure internal Dict constructor is called with the correct arguments

Seems to have been added by surprise in 20af397357f181291d3267163536e945383948af with no users or tests.

vtjnash commented:
JLD2 is likely the result of a bug (https://github.com/JuliaIO/JLD2.jl/pull/492). I am slowly working through the list of other failures, though many of them are a result of package bugs, either because they rely on implementation details (making synthetic calls to `:new` that are not well-formed here) or because they attempt to violate the object model with calls to `pointer` and `pointer_from_objref` or they attempt to type-pirate (defining methods on getproperty of Array)

A lot of them are due to https://github.com/JuliaIO/PNGFiles.jl/pull/69 (and sometimes other similar ones like https://github.com/evetion/GeoAcceleratedArrays.jl/pull/4)

vtjnash commented:
@nanosoldier `runtests(["SHTOOLS", "Elliptic", "TrueSkillThroughTime", "ConstructionBase", "Flatten", "PhaseRec", "Materials", "Determinantal", "ChunkedBase", "NaNStatistics", "JLD2", "HomalgProject", "NgSpice", "TranscodingStreams", "TimeSeriesClassification", "LazyAlgebra", "JsonGrinder", "ImageIO", "SparseKmeansFeatureRanking", "OrdinalGWAS", "SentinelArrays", "OSMMakie", "Krylov", "SMLMData", "ModeCouplingTheory", "MadNLP", "ReactionSensitivity", "Undefs", "GeNIOS", "Distances", "Fronts", "PeriodicSchurDecompositions", "FiniteDiff", "Muspel", "GLPK", "RuntimeGeneratedFunctions", "AffineMaps", "IRKGaussLegendre", "RData", "LinearNoiseApproximation", "SimpleConfig", "BioFetch", "Sundials", "MomentClosure", "ComplexMixtures", "StrideArraysCore", "NaiveNASlib", "YAML", "StaticArrayInterface", "VideoIO", "ConstraintSolver", "SparseIR", "LatticeQCD", "Demes", "DynamicSparseArrays", "TrajOptPlots", "PPTX", "LibModbus", "Attractors", "LightOSM", "Resizing", "DynamicAxisWarping", "ChaosTools", "GPUArrays", "Alexya", "FastRunningMedian", "GenomicAnnotations", "Consensus", "PNGFiles", "Fenrir", "AbstractAlgebra", "WordCloud", "EMpht", "CachedFunctions", "PowerFlows", "LocalPoly", "DynamicPolynomials", "ImageMorphology", "PALEOboxes", "ProtoBuf", "StationXML", "PowerSystemsMaps", "PowerNetworkMatrices", "Sunny", "SeisPDF", "PowerPlots", "CxxWrap", "GeneticsMakie", "MosekTools", "MIRTio", "GMT", "GeoAcceleratedArrays", "SurfaceCoverage", "FullNetworkSystems", "CCBlade", "HurdleDMR", "Gaius", "PRONTO", "Bobby", "Singular", "MutableArithmetics", "ArviZPythonPlots", "DelaySSAToolkit", "DataDrivenAcoustics", "DataFlowTasks", "ManifoldsBase", "TypedPolynomials", "FractalDimensions", "SMLMFrameConnection", "Strided", "BandedMatrices", "StochasticDelayDiffEq", "ParametricOptInterface", "NaiveNASflux", "EpithelialDynamics1D", "DiffusionMap", "HighVoronoi", "LinearMaps", "ClosedLoopReachability", "PerfChecker", "HmtArchive", "ModelPredictiveControl", "MRINavigator", "ManifoldDiff", "DifferentiableBackwardEuler"])`

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51319#issuecomment-1768927027) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8309b27_vs_e36f65f/report.html) is available.

LilithHafner commented on base/Base.jl:
```diff
@@ -6,10 +6,19 @@ using Core.Intrinsics, Core.IR
 # that doesn't require any function (except what we can get from the `Core` top-module)
-const _included_files = Array{Tuple{Module,String},1}(Core.undef, 1)
+# start this big so that we don't have to resize before we have defined how to grow an array
+const _included_files = Array{Tuple{Module,String},1}(Core.undef, 50)
```
```suggestion
const _included_files = Array{Tuple{Module,String},1}(Core.undef, 400)
```
For performance reasons :P

But also in case we learn how to grow an array later.

oscardssmith committed:
Update base/Base.jl

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

vtjnash committed:
some fixes

oscardssmith commented:
This is now ready to merge once CI is happy with it.

Commit:
fix typo

LilithHafner commented:
14b32185d65c77bd51ebf6ad9200ed0aebadcfb3 may be the first win for #51704!

vchuravy commented:
Congratulations 🎉🎉

fingolfin commented:
Seems like a great improvement, but unfortunately it also broke the Julia C API. As a consequence CxxWrap (and probably a few other things) don't work anymore and need to be fixed. @benlorenz started to work on fixing it, see <https://github.com/JuliaInterop/libcxxwrap-julia/pull/137>, but it's not yet working.

To a degree, I understand why API breakage keeps happening, but it'd be nice if we could slowly over time start to stabilize some APIs and make some promises about keeping them around a bit longer -- APIs for accessing arrays seem like a prime candidate for that.

Otherwise, certain packages which are not *very* actively maintained will just rot away over time, I am afraid.
